### PR TITLE
Apply 'convention over configuration' to lex/parse API.

### LINKED ISSUE #22: Refactor parse/lex/eval to use a very simple API
parse() should take a string and do parsing, with no explicit lexer, mocks, and with optimize=1

### PR DESCRIPTION
**Pull request by Renelvon**
### Lex:
- PLY arguments are now explicitly stated (and part of interface).
- Build own `logger` if one is not provided.
- Rename `_LexerBuilder` to `_LexerFactory` to follow semantics.
- Add `tokenize` function at module level to enable hassle-free lexing.
### Parse:
- Silence PLY errors/warnings when building a parser with non-default initial state.
- Build own `logger` if one is not provided.
- Build own `lexer` if one is not provided.
- Add `parse` function at module level to enable hassle-free parsing.
### Tests:
- Add tests for module-level functions.
- Use the default objects/arguments of the new API.
- Do some DRY & pep8-ing.

Fix #22.

### CODE REVIEW

File: tests/test_lexer.py [hunk 1]
```diff
@@ -25,17 +25,11 @@ def _assert_individual_token(self, input, expected_type, expected_value):
...
-        list(lexer)  # Force lexing
+        l, mock = self._lex_data(input)
         mock.success.should.be.ok
 
     def _assert_lex_failure(self, input):
-        mock = error.LoggerMock()
-        lexer = lex.Lexer(logger=mock)
-        lexer.input(input)
-        list(lexer)  # Force lexing
+        l, mock = self._lex_data(input)
```
Comment (dionyziz) [inline tests/test_lexer.py:29]: Use `tokenize` in implementation of `._lex_data`

File: tests/test_lexer.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
```
Comment (dionyziz) [inline tests/test_lexer.py:4]: Add tests for `tokenize`

File: tests/test_parser.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
```
Comment (dionyziz) [inline tests/test_parser.py:2]: Unit test module-level `parse`

File: tests/test_type.py [hunk 4]
```diff
@@ -64,9 +64,8 @@ def test_function_types(self):
 
     def _process_typedef(self, typeDefListList):
         mock = error.LoggerMock()
-        lexer = lex.Lexer(logger=mock, optimize=1)
-        parser = parse.Parser(logger=mock, optimize=1)
-        parser.parse(data=typeDefListList, lexer=lexer)
+        parser = parse.Parser(logger=mock)
+        parser.parse(data=typeDefListList)
```
Comment (dionyziz) [inline tests/test_type.py]: Use `parse.parse`

File: tests/test_parser.py [hunk 5]
```diff
@@ -16,23 +16,17 @@ def setUp(self):
...
-                debug=0
+                optimize=False,
+                start=start
             )
 
-        tree = parser.parse(
-            data=data,
-            lexer=lexer
-        )
+        tree = parser.parse(data=data)
```
Comment (dionyziz) [inline tests/test_parser.py:29]: Use `parse.parse`

File: tests/test_type.py [hunk 6]
```diff
@@ -64,9 +64,8 @@ def test_function_types(self):
...
         return parser.logger.success
```
Comment (dionyziz) [inline tests/test_type.py]: TBD: `parser.logger.success` VS `mock.success`

File: tests/test_lexer.py (hunk 2 above)
Comment (Renelvon) [inline tests/test_lexer.py:4]: Done. Just an afterthought: If I use the `tokenize` inside `_lex_data` isn't that enough testing?

File: tests/test_lexer.py (hunk 1 above)
Comment (Renelvon) [inline tests/test_lexer.py:29]: Done.

File: tests/test_parser.py (hunk 3 above)
Comment (Renelvon) [inline tests/test_parser.py:2]: Done.

File: tests/test_type.py (hunk 6 above)
Comment (Renelvon) [inline tests/test_type.py]: If we are using `parse.parse`, then there is no relevant `parser` to query. We can only use `mock.success`.

File: tests/test_type.py (hunk 4 above)
Comment (Renelvon) [inline tests/test_type.py]: Done.

File: tests/test_parser.py (hunk 5 above)
Comment (Renelvon) [inline tests/test_parser.py:29]: And forego all that nice memoization?

File: tests/test_lexer.py (hunk 2 above)
Comment (dionyziz) [inline tests/test_lexer.py:4]: Thanks :)

No. If `tokenize` and the tests invoking `_lex_data` are both wrong, the tests may pass even through the code does not follow the contract.

File: tests/test_parser.py (hunk 5 above)
Comment (dionyziz) [inline tests/test_parser.py:29]: You're right, never mind.

**Pull request merged**